### DISCUSSION TITLE
Sanity Check: Triggering pgsql-http post on auth.users update

### DISCUSSION THREAD

**Discussion by tourbillonlabs** [supabase/supabase#4143] [Questions] [answered]
Hi all,

Just need a sanity check on adding a user email as a subscriber on a service like mailchimp after a user's first login.

1. User creates account via social or email, they are sent the Supabase welcome email. For email users, they are asked to verify their email address.
2. Once they verify/log in for the first time, I want to trigger adding them to the subscriber list of my email marketing sending service.

I'm not doing this client-side/with supabase-js because I can't see a way to determine if it's a user's first login from the available interfaces. That said, if there's a way to do this, would love to hear it.

The remaining option appears to be using a postgres trigger on the auth.users table.

I have the following:

```
CREATE OR REPLACE FUNCTION auth.add_subscriber()
    RETURNS TRIGGER AS $$
DECLARE
    payload TEXT;
BEGIN
    IF (OLD IS NULL OR OLD.last_sign_in_at IS NULL) AND NEW.last_sign_in_at IS NOT NULL THEN
        payload := json_build_object(
            'email', NEW.email
        );
        perform extensions.http((
            'POST',
            THE_ENDPOINT,
            ARRAY[extensions.http_header('THE_API_KEY_HEADER', 'THE_API_KEY')],
            'application/json',
            payload::varchar
        ));
    END IF;
    RETURN NEW;
END;
$$ LANGUAGE plpgsql SECURITY DEFINER;

CREATE TRIGGER on_auth_user_created_add_subscriber
AFTER INSERT OR UPDATE ON auth.users
FOR EACH ROW EXECUTE PROCEDURE auth.add_subscriber();
```

When I click on the verify email link to trigger this, it falls over at the `http_post` call and I get the error:

```
{"code":500,"msg":"Database error granting user","error_id":"bd2183b0-b6ff-4afc-abd6-1a5bc95e8bd9"}
```

But I can't figure out what the issue is. Omitting the http post call allows the verification to complete successfully. I've tried creating the procedure under the auth and public schemas, and executing the post by itself works fine.

Should this work, or have I missed something critical?

Thanks

Edit: specified `extensions` for the `http_header` as well.

**Comment by GaryAustin1** [comment]
I've been messing around with search paths and extensions today.  When I use the http function I have the search path set in the security definer function to include extensions.  I notice you hard code extensions.http( ...   
Is http_header also part of that extension and would it need extensions.http_header( ?  or does using it in the http function make it "known".  
My guess is it needs a schema or path like http() as an auth trigger does not have extensions in the path (I don't believe).

**Reply by tourbillonlabs** [reply]
Gosh, that would have been embarrassing if that was the problem. Fortunately/Unfortunately, I just tried specifying `extensions.http_header` and it still errors out. :(

Thanks for pointing that out though!

**Reply by GaryAustin1** [reply]
OK, not sure how this would work if you tested http standalone, but don't you need http((stuff)::http_request).  I don't see http_request.  Maybe it is optional, but it is in my code and the example.
I would also use set search_path = public,auth,extensions; (or at least the extensions part) in the function just in case.

**Reply by tourbillonlabs** [reply]
I think the casting is optional. I did have it there originally. And I just tried it with

```
        perform extensions.http((
            'POST',
            THE_ENDPOINT,
            ARRAY[extensions.http_header('THE_API_KEY_HEADER', 'THE_API_KEY')],
            'application/json',
            payload::varchar
        )::extensions.http_request);
```

and it still returns:

```
{"code":500,"msg":"Database error granting user","error_id":"542cf37d-3455-4aca-a2c6-3a52da39ea54"}
```

**Reply by tourbillonlabs** [reply]
Oh. That's new. It's actually firing off the call successfully. My mailing service has added the email.

But I'm still getting the `Database error granting user` error returned from the supabase email verification link. Maybe the `extensions.http_request` isn't optional after all. (Though I definitely did successfully invoke it standalone)

This is progress, though I guess that means the error is a separate problem?

**Reply by GaryAustin1** [reply]
Edit: first part of this was before http started working.
At this point, if it were me, I'd do my auth trigger/function to update a column on my profiles table and trigger http function off that with a trigger function to eliminate the auth trigger being unique environment.
Edit.  I see you have it working better.  Your error is just generic for something did not work in the entire auth trigger process.

**Reply by tourbillonlabs** [reply]
Good to know about the generic error. Shame it isn't less generic. :)

I don't need the email in my profiles table at all, but if that's what I need to do to both access the email and complete the trigger action without it bailing out on the auth trigger, I'll take it. Will give it a go tomorrow and report back.

Thanks for the sanity check and guidance!

**Reply by GaryAustin1** [reply]
**Edit: there is an audit log in auth settings of the ui that MIGHT have more info on the error.**
Something  else to think about using is pg_net possibly.  It is async (versus http which locks up until your email service responds).  I can't use it because I use delete in my http call and it only supports post and get.  It is what the Supabase function hooks use.  No idea if delaying auth process too long could cause an issue or how easy to use.
https://supabase.github.io/pg_net/api/
edit: appears gotrue has a 3 second default (no idea what supabase sets it to) timeout for gotrue http call to the database and then to get your http call out and all back.  Seems long....

**Comment by GaryAustin1** [comment, marked as answer]
At this point, if it is true that GOTRUE_WEBHOOK_TIMEOUT_SEC=3 is used in their call to the database (I believe it is, as it uses http()), then your approach is flawed in general (possibly even without that limitation on a hang of the http()).  You will fail a user signing up if your email service is down and delays response at a minimum.  So you have to put some timeout logic in your code (is that possible?) or use pg_net to go with call in the auth trigger path.  Note that even writing to another table and trigger off of that does not solve it as still in same over call from gotrue.

You might consider going back to client side approach with a local storage variable that gets set to not call a setup rpc function again and then have intelligence in your database to check if you already sent email to your service (with a table to track that).  Or do some sort of cron sweep on a table of new emails from auth then clear it.

NOTE: if you continue to want to try http() call I really believe you need to set the search path in your function.  The http code has this:
_CREATE OR REPLACE FUNCTION http(request http_request)
    RETURNS **http_response**
    AS 'MODULE_PATHNAME', 'http_request'
    LANGUAGE 'c';_
where http_response probably also fails without extensions in the path after it sends the network request out....

**Reply by tourbillonlabs** [reply]
After you mentioned the timeout I realised the flaw as well. I've decided to stick with my client-side solution that on auth, checks for a flag in the profile before attempting to add them as a subscriber, and sets the flag accordingly. At least this means I won't interfere with the auth triggers.

The info about gotrue, and the http and net extensions has been very informative though, and I appreciate you working this out with me. :)

Thanks!

**Comment by saendu** [comment]
For anyone that ends up here. I had the same issue that a trigger function that was containing a http_post threw a 500 error everytime it was executed. Removing the http_post from the trigger function, made the function work. So it was for sure related.  

As @GaryAustin1 suggested, adding the line ``search_path = public,auth,extensions`` resolved my problem. 
```
...
security definer set search_path = public,auth,extensions
...
```

Also make sure that you store the returning content from the http_post to a variable with ``SELECT status, content::varchar into res`` else you will face the error of not using your result.

Hope this helps.
